### PR TITLE
Chapter 7

### PR DESCRIPTION
**Pull request by Arizar98**

### CODE REVIEW

**Comment by Arizar98** [comment]
Ready

File: app/assets/stylesheets/custom.scss [hunk 1]
```diff
@@ -32,3 +32,86 @@ textarea {
...
+    margin-top: 20px;
+    &:first-child {
+      border: 0;
+      padding-top: 0;
+    }
+  span {
+    display: block;
+    margin-bottom: 3px;
+    line-height: 1;
+  }
```
Comment (saobv-2223) [inline app/assets/stylesheets/custom.scss]: image

File: app/assets/stylesheets/users.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Users controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
Comment (saobv-2223) [inline app/assets/stylesheets/users.scss]: remove file that never used

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
+    if @user.save
+      flash[:success] = t "welcome_to_the_sample_app!"
+      redirect_to @user
+    else
+      render :new
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb:19]: add flash for this case

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  def show
+    @user = User.find_by id: params[:id]
+  end
+
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: remove ()

File: app/views/users/new.html.erb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    <%= f.label :name %>
+    <%= f.text_field :name %>
+    <%= f.label :email %>
+    <%= f.email_field :email %>
+    <%= f.label :password %>
+    <%= f.password_field :password %>
+    <%= f.label :password_confirmation, t("confirmation") %>
+    <%= f.password_field :password_confirmation %>
+    <%= f.submit t("create_my_account"), class: "btn btn-primary" %>
+  <% end %>
```
Comment (saobv-2223) [inline app/views/users/new.html.erb]: add 2 spaces

File: app/views/users/show.html.erb [hunk 6]
```diff
@@ -0,0 +1,14 @@
+
```
Comment (saobv-2223) [inline app/views/users/show.html.erb]: remove line 1

File: app/views/users/show.html.erb [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+<% provide :title, @user.name %>
+<div class="row">
+  <aside class="col-md-4">
+  <section class="user_info">
+  <h1>
+    <%= gravatar_for @user %>
+    <%= @user.name %>
+  </h1>
+  </section>
```
Comment (saobv-2223) [inline app/views/users/show.html.erb]: indenting 2 spaces

File: config/routes.rb [hunk 8]
```diff
@@ -1,7 +1,13 @@
 Rails.application.routes.draw do
+  get 'users/show'
+  get 'users/new'
```
Comment (saobv-2223) [inline config/routes.rb]: using " and move into scope

**Comment by Arizar98** [comment]
Ready

File: app/assets/stylesheets/custom.scss [hunk 9]
```diff
@@ -20,15 +20,133 @@ textarea {
...
+  span {
+    display: block;
+
+    margin-bottom: 3px;
+
+    line-height: 1;
+  }
+  h1 {
+    font-size: 1.4em;
+
```
Comment (saobv-2223) [inline app/assets/stylesheets/custom.scss]: check indenting again

File: app/controllers/users_controller.rb [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+      flash[:success] = t "welcome_to_the_sample_app!"
+      redirect_to @user
+    else
+      render :new
+      flash[:alert] = t ".try_again"
+    end
+  end
+
+  private
+  def user_params
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]:
```suggestion
  private
  
  def user_params
```

File: app/views/users/new.html.erb [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    <%= f.label :password %>
+
+    <%= f.password_field :password %>
+
+    <%= f.label :password_confirmation, t("confirmation") %>
+
+    <%= f.password_field :password_confirmation %>
+
+    <%= f.submit t("create_my_account"), class: "btn btn-primary" %>
+  <% end %>
```
Comment (saobv-2223) [inline app/views/users/new.html.erb]: indenting 2 spaces for those lines

File: app/views/users/show.html.erb [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+
+<% provide :title, @user.name %>
+<div class="row">
+  <aside class="col-md-4">
+  <section class="user_info">
+    <h1>
+      <%= gravatar_for @user %>
+      <%= @user.name %>
+    </h1>
+  </section>
```
Comment (saobv-2223) [inline app/views/users/show.html.erb]: indenting 2 spaces for those lines

**Comment by Arizar98** [comment]
Ready

File: app/assets/stylesheets/custom.scss [hunk 13]
```diff
@@ -20,15 +20,133 @@ textarea {
...
   color: $logo_colour;
+
   text-transform: uppercase;
+
   letter-spacing: -1px;
+
   padding-top: 9px;
+
   font-weight: bold;
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]: remove empty lines

File: app/assets/stylesheets/custom.scss [hunk 14]
```diff
@@ -20,15 +20,133 @@ textarea {
...
 }
+
 #logo:hover {
   color: $logo_colour;
+
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]: remove this line

File: app/assets/stylesheets/custom.scss [hunk 15]
```diff
@@ -20,15 +20,133 @@ textarea {
...
+@mixin box_sizing {
+  -moz-box-sizing:
+
+  border-box;
+
+  -webkit-box-sizing: border-box;
+
+  box-sizing:
+
+  border-box;
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]: remove empty lines

File: app/assets/stylesheets/custom.scss [hunk 16]
```diff
@@ -20,15 +20,133 @@ textarea {
...
+.debug_dump {
+  clear: both;
+
+  float: left;
+
+  width: 100%;
+
+  margin-top: 45px;
+
+  @include box_sizing;
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]: remove empty lines

File: app/assets/stylesheets/custom.scss [hunk 17]
```diff
@@ -20,15 +20,133 @@ textarea {
...
+}
+
+
+/* sidebar */
+aside {
+  section.user_info {
+    margin-top: 20px;
+  }
+  section {
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]:
```suggestion
  }
  
  section {
```

File: app/assets/stylesheets/custom.scss [hunk 18]
```diff
@@ -20,15 +20,133 @@ textarea {
...
+    padding: 10px 0;
+
+    margin-top: 20px;
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]:
```suggestion
    padding: 10px 0;
    margin-top: 20px;
```

File: app/assets/stylesheets/custom.scss [hunk 19]
```diff
@@ -20,15 +20,133 @@ textarea {
...
+
+    &:first-child {
+    border: 0;
+
+    padding-top: 0;
+    }
+  span {
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]:
```suggestion
    border: 0;
    padding-top: 0;
    }
    
  span {
```

File: app/assets/stylesheets/custom.scss [hunk 20]
```diff
@@ -20,15 +20,133 @@ textarea {
...
+    display: block;
+
+    margin-bottom: 3px;
+
+    line-height: 1;
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]:
```suggestion
    display: block;
    margin-bottom: 3px;
    line-height: 1;
```

File: app/assets/stylesheets/custom.scss [hunk 21]
```diff
@@ -20,15 +20,133 @@ textarea {
...
+
+    text-align: left;
+
+    letter-spacing: -1px;
+
+    margin-bottom: 3px;
+
+    margin-top: 0px;
+  }
+  }
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]:
```suggestion
  }
  
  h1 {
    font-size: 1.4em;
    text-align: left;
    letter-spacing: -1px;
    margin-bottom: 3px;
    margin-top: 0px;
  }
```

File: app/assets/stylesheets/custom.scss [hunk 22]
```diff
@@ -20,15 +20,133 @@ textarea {
...
+}
+
+.gravatar {
+  float: left;
+
+  margin-right: 10px;
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]:
```suggestion
  float: left;
  margin-right: 10px;
```

File: app/assets/stylesheets/custom.scss [hunk 23]
```diff
@@ -20,15 +20,133 @@ textarea {
...
+
+/* forms */
+input, textarea, select, .uneditable-input {
+  border: 1px solid #bbb;
+
+  width: 100%;
+
+  margin-bottom: 15px;
+
+  @include box_sizing;
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]:
```suggestion
  border: 1px solid #bbb;
  width: 100%;
  margin-bottom: 15px;
  @include box_sizing;
```

File: app/controllers/users_controller.rb [hunk 24]
```diff
@@ -0,0 +1,27 @@
...
+  end
+
+  def create
+    @user = User.new user_params
+    if @user.save
+      flash[:success] = t "welcome_to_the_sample_app!"
+      redirect_to @user
+    else
+      render :new
+      flash[:alert] = t ".try_again"
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]:
```suggestion
      flash.now[:alert] = t ".try_again"
```

File: app/models/user.rb [hunk 25]
```diff
@@ -1,6 +1,6 @@
 class User < ApplicationRecord
 
-  before_save {self.email = email.downcase}
+  before_save{self.email = email.downcase}
```
Comment (chungpv-1008) [inline app/models/user.rb]:
```suggestion
  before_save{email.downcase!}
```

File: app/views/layouts/application.html.erb [hunk 26]
```diff
@@ -12,9 +12,13 @@
   <body>
   <%= render "layouts/header" %>
   <div class="container">
+    <% flash.each do |message_type, message| %>
+      <div class="alert alert-<%= message_type %>"><%= message %></div>
+    <% end %>
     <%= yield %>
+
+    <%= debug(params) if Rails.env.development? %>
```
Comment (chungpv-1008) [inline app/views/layouts/application.html.erb]:
```suggestion
    <%= debug params if Rails.env.development? %>
```

File: app/views/users/new.html.erb [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+<div class="row">
+ <div class="col-md-6 col-md-offset-3">
+  <%= form_for @user, url: signup_path do |f| %>
+
+   <% if @user.errors.any? %>
+    <div id="error_explanation">
+      <div class="alert alert-danger">
+        <%= t "the_form_contains" %>
+
+        <%= pluralize @user.errors.count, "error" %>.
```
Comment (chungpv-1008) [inline app/views/users/new.html.erb]:
```suggestion
        <%= pluralize @user.errors.size, "error" %>.
```

File: config/routes.rb [hunk 28]
```diff
@@ -1,7 +1,14 @@
 Rails.application.routes.draw do
   scope "(:locale)", locale: /en/ do
+    get "users/show"
+    get "users/new"
+
```
Comment (chungpv-1008) [inline config/routes.rb]:
```suggestion
```

File: config/routes.rb [hunk 29]
```diff
@@ -1,7 +1,14 @@
...
     root "static_pages#home"
     get "static_pages/help"
     get "static_pages/contact"
+
+    get "/signup", to: "users#new"
+    resources :users
```
Comment (chungpv-1008) [inline config/routes.rb]:
```suggestion
    resources :users, only: %i(show new)
```

**Comment by Arizar98** [comment]
Ready

**Comment by manhnt-0960** [comment]
commented
please merge 2 commits to 1 commit

**Comment by manhnt-0960** [comment]
commented
please merge 2 commits to 1

**Comment by Arizar98** [comment]
Ready

File: app/controllers/users_controller.rb [hunk 30]
```diff
@@ -0,0 +1,27 @@
+class UsersController < ApplicationController
+  def show
+    @user = User.find_by id: params[:id]
```
Comment (vanvtt-0952) [inline app/controllers/users_controller.rb:3]: bỏ thêm code xử lý khi nil nhé

`return if @user`
`flash[:warning] = "User not found"`
`redirect_to root_path`

File: app/controllers/users_controller.rb (hunk 24 above)
Comment (vanvtt-0952) [inline app/controllers/users_controller.rb]: ` flash[:alert] = t ".try_again"` thành ` flash.now[:alert] = t ".try_again"`

File: app/views/users/new.html.erb [hunk 31]
```diff
@@ -0,0 +1,41 @@
...
+
+ <%= f.label :password %>
+
+ <%= f.password_field :password %>
+
+ <%= f.label :password_confirmation, t("confirmation") %>
+
+ <%= f.password_field :password_confirmation %>
+
+ <%= f.submit t("create_my_account"), class: "btn btn-primary" %>
```
Comment (vanvtt-0952) [inline app/views/users/new.html.erb]: indent incorrect

File: config/routes.rb [hunk 32]
```diff
@@ -1,7 +1,13 @@
 Rails.application.routes.draw do
   scope "(:locale)", locale: /en/ do
+    get "users/show"
+    get "users/new"
```
Comment (vanvtt-0952) [inline config/routes.rb]: you can remote line number 3 and 4

Comment (Arizar98) [reply config/routes.rb]: I'm sorry, can you please explain more, I don't understand what I should do here (^^;)

**Comment by Arizar98** [comment]
Ready

File: app/assets/stylesheets/custom.scss [hunk 33]
```diff
@@ -28,7 +28,96 @@ textarea {
...
+/* sidebar */
+aside {
+  section.user_info {
+    margin-top: 20px;
+  }
+
+  section {
+    padding: 10px 0;
+    margin-top: 20px;
+    &:first-child {
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]:
```suggestion
    margin-top: 20px;
    
    &:first-child {
```

File: app/assets/stylesheets/custom.scss [hunk 34]
```diff
@@ -28,7 +28,96 @@ textarea {
...
+    border: 0;
+    padding-top: 0;
+    }
+  span {
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]:
```suggestion
    }
    
  span {
```

File: app/assets/stylesheets/custom.scss [hunk 35]
```diff
@@ -28,7 +28,96 @@ textarea {
...
+    display: block;
+    margin-bottom: 3px;
+    line-height: 1;
+  }
+  h1 {
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]:
```suggestion
  }
  
  h1 {
```

File: app/assets/stylesheets/custom.scss [hunk 36]
```diff
@@ -28,7 +28,96 @@ textarea {
...
+    font-size: 1.4em;
+    text-align: left;
+    letter-spacing: -1px;
+    margin-bottom: 3px;
+    margin-top: 0px;
+  }
+  }
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]:
```suggestion
  }
```

File: app/assets/stylesheets/custom.scss [hunk 37]
```diff
@@ -28,7 +28,96 @@ textarea {
...
+}
+
+.gravatar_edit {
+  margin-top: 15px;
+}
+
+
+/* forms */
+input, textarea, select, .uneditable-input {
+  border: 1px solid #bbb;
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]: color

File: config/routes.rb (hunk 32 above)
Comment (chungpv-1008) [inline config/routes.rb]: remove

File: config/routes.rb [hunk 38]
```diff
@@ -1,7 +1,13 @@
...
     root "static_pages#home"
     get "static_pages/help"
     get "static_pages/contact"
+
+    get "/signup", to: "users#new"
+    resources :users, only: %i(show new)
+    post "/signup", to: "users#create"
```
Comment (chungpv-1008) [inline config/routes.rb]:
```suggestion
    get "/signup", to: "users#new"
    post "/signup", to: "users#create"
    
    resources :users, only: %i(show new)
```

**Comment by Arizar98** [comment]
Ready

**Comment by ngocvt-0484** [comment]
merged

**Pull request merged**